Fix bug in compaction props

ddanielr opened the pull request:
Fixes an issue with the deprecated `tserver.compaction.major.thread.files.open.max` property. 

If the `open.max` property was set, the replacement property `tserver.compaction.major.service.default.planner.opts.maxOpen` would not be used and any user-defined compaction service would also only use the old `open.max` property value. 

The cause was an error in the `getFullyQualifiedOption` of `CompactionPlannerInitParams`. 

This has now been fixed and a test included to reproduce the problematic behavior.

a268958ddf411e460e85b9eb2cf71ab5fee5a4cd will fail the tests, and 4042529ac0491f0d25b7dc0f6cc89d547a1245cc includes the one line fix.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlanner.java:
```diff
@@ -225,6 +225,10 @@ private void determineMaxFilesToCompact(InitParameters params) {
     }
   }
 
+  public int getMaxFilesToCompact() {
+    return this.maxFilesToCompact;
+  }
```
Is this SPI addition necessary to implement the bug fix? It would good to try to avoid the change in a user-facing component in a bug fix release. We can be looser about it with SPI rather than public API, if necessary, but if it's not necessary, it'd be good to keep it the same.

ddanielr replied:
I refactored my test to allow for a compaction example that does not require the accessor method. 
There are now, no changes to DefaultCompactionPlanner.java

keith-turner commented on core/src/test/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlannerTest.java:
```diff
@@ -239,6 +243,38 @@ public void testMaxSize() {
+
+    // Test with max.open property
+    planner.init(getInitParams(senv, executors, "15"));
+    assertEquals(5, planner.getMaxFilesToCompact());
```
If this works it would test a bit more and would not need to add the getMaxFilesToCompact() method

```suggestion
    var all = createCFs("F1", "1M", "F2", "1M", "F3", "1M", "F4", "1M", "F5", "11M","F6","1M","F7","1M");
    var params = createPlanningParams(all, all, Set.of(), 2, CompactionKind.SYSTEM);
    var plan = planner.makePlan(params);
    // TODO verify job has 5 files
```

ddanielr replied:
I ended up refactoring the compaction tests in DefaultCompactionPlannerTest to use a default conf and then set the property prior to compacting for tests that checked the plan files.

This cuts down on the amount of custom test code we were using while also getting these tests closer to mimicking a real-world situation.

keith-turner replied:
> while also getting these tests closer to mimicking a real-world situation.

That is nice.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlannerTest.java:
```diff
@@ -239,6 +243,38 @@ public void testMaxSize() {
+        maxOpen.toString());
+    planner = new DefaultCompactionPlanner();
+    planner.init(getInitParams(senv, executors, maxOpen.toString()));
+    assertEquals(maxOpen, planner.getMaxFilesToCompact());
```
Could set max open to 10 and pass in 11 files of all the same size  like the prev comment and verify it creates a job of size 10.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlannerTest.java:
```diff
@@ -486,80 +485,36 @@ public Builder createPlanBuilder() {
+    String maxOpen;
+    maxOpen =
+        conf.get(Property.TSERV_COMPACTION_SERVICE_PREFIX.getKey() + "cs1.planner.opts.maxOpen");
+    if (maxOpen == null) {
+      if (conf.isSet(deprecatedMaxOpen.getKey())) {
+        maxOpen = conf.get(deprecatedMaxOpen.getKey());
+      } else {
+        maxOpen = Property.TSERV_COMPACTION_SERVICE_DEFAULT_MAX_OPEN.getDefaultValue();
```
This feels like its duplicating code in the Defaultcompactionplanner, but not sure.  Wondering if it could be removed.  If not why is it needed?


```suggestion
    if (maxOpen == null) {
        maxOpen = Property.TSERV_COMPACTION_SERVICE_DEFAULT_MAX_OPEN.getDefaultValue();
```

ddanielr replied:
While the DefaultCompactionPlanner code looks to see if the `maxOpen` property is set, it never uses that value and instead checks the `options` for the `maxOpen` vialue.  
```
this.maxFilesToCompact = Integer.parseInt(params.getOptions().getOrDefault("maxOpen", "10"));
``` 
So, for our test we have to include the property in the conf and also in the `options` map since that is where the property values would eventually end up after being parsed by CompactionServicesConfig. 

Since you pointed that out, I'm going to change the logic in determineMaxFilesToCompact to make this a bit simpler.

ddanielr replied:
@keith-turner brought up a good point with the default compaction services and the use of ConfigurationCopy. 
I'll need to make some additional changes to this PR so the default compaction services accept the deprecated property behavior correctly.

ddanielr commented on core/src/main/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlanner.java:
```diff
@@ -212,16 +212,16 @@ public void init(InitParameters params) {
+        .isSet(Property.TSERV_MAJC_THREAD_MAXOPEN.getKey())) {
       log.warn("The property " + Property.TSERV_MAJC_THREAD_MAXOPEN.getKey()
-          + " was set, it is deprecated.  Set the " + fqo + " option instead.");
+          + " was set, it is deprecated.  Set the " + params.getFullyQualifiedOption("maxOpen")
```
We only have to check if the deprecated property is set when no valid planner option exists. 

The `maxOpen` property is already parsed out by the CompactionServicesConfig class and included in the DefaultCompactionPlanner's options.

The only time the "fully qualified"  method should be used is when informing the user to update a specific property in their config.

ddanielr commented:
Switching this to draft until the proper changes are made for the defaultCompactionService behaviors

ddanielr commented:
Removed some earlier changes and added test case for the default compaction service to ensure that user set properties correctly overrode default values.

Now this PR does not change the property lookup logic in the DefaultCompactionPlanner. It does replace a hardcoded maxOpen value with a default value reference,

keith-turner commented on core/src/main/java/org/apache/accumulo/core/conf/ConfigurationCopy.java:
```diff
@@ -30,6 +30,8 @@
  */
 public class ConfigurationCopy extends AccumuloConfiguration {
+
+  AccumuloConfiguration parent = null;
```
Would be good to do something similar to this for isSet.  Like the way it all checks the parent.  I don't think it matters for this PR, but would make the use parent consistent in case this is used by other stuff.

https://github.com/apache/accumulo/blob/2890cc2c99a1ae0fd1531533df6ee75f8b2f1cfa/core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java#L249

ddanielr replied:
Created https://github.com/apache/accumulo/pull/4112 to update that method.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlannerTest.java:
```diff
@@ -152,12 +171,126 @@ public void testRunningCompaction() {
+    EasyMock.replay(senv);
+
+    // Use the CompactionServicesConfig to create options based on default property values
+    var CompactionServices = new CompactionServicesConfig(aconf, log::warn);
```
This variable name was confusing me, I thought it was a class name later and thought something was calling static method on a class.

```suggestion
    var compactionServices = new CompactionServicesConfig(aconf, log::warn);
```

ddanielr replied:
Yeah that's an oversight. I'll fix the var naming.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/spi/compaction/DefaultCompactionPlannerTest.java:
```diff
@@ -152,12 +171,126 @@ public void testRunningCompaction() {
+  @Test
+  @SuppressWarnings("removal")
+  public void testOverrideMaxOpen() {
+    ConfigurationCopy aconf = new ConfigurationCopy(DefaultConfiguration.getInstance());
```
Can make the default config the parent which make the test a bit more like the real env.

```suggestion
    ConfigurationCopy aconf = new ConfigurationCopy(Map.of(), DefaultConfiguration.getInstance());
```

ctubbsii commented:
I reverted this in d4a7a8984305306f3f0a4c594f64634cd65114a9 in favor of #4117, which keeps all the changes in this #4092 but uses SiteConfiguration instead of ConfigurationCopy to test overrides of system defaults, but without #4112 (which is OBE without the changes to ConfigurationCopy).
